File: bms/go/go-gt/localSearch/localSearch.go
package test

import "math/rand"

var ()

const ()

type Vector []int64

func localSearch(a *Matrix, b *Matrix, p Vector, cost *int64) {
	// set of moves, numbered from 0 to index
	var i, j, nMov int64
	n := p.Len()
	move := make(Vector, n*(n-1)/2)
	nMov = 0
	for i = 0; i < n-1; i++ {
		for j = i + 1; j < n; j++ {
			move[nMov] = n*i + j
			nMov++
		}
	}
	improved := true
	for k := 0; k < 2 && improved; k++ {
		improved = false
		for i = 0; i < nMov-1; i++ {
			move.Swap(i, unif(i+1, nMov-1))
		}
		for i = 0; i < nMov; i++ {
			r := move[i] / n
			s := move[i] % n
			d := delta(a, b, p, r, s)
			if d < 0 {
				*cost += d
				p.Swap(r, s)
				improved = true
			}
		}
	}
	return
}

func (v Vector) Len() int64 {
	return int64(len(v))
}

func unif(low, high int64) int64 {
	return low + int64(float64(high-low+1)*rand.Float64())
}

func (p Vector) Swap(i int64, j int64) {
	x := p[i]
	p[i] = p[j]
	p[j] = x
}

func delta(a *Matrix, b *Matrix, p Vector, r int64, s int64) (d int64) {
	var i int64
	d = int64((a.Get(r, r)-a.Get(s, s))*(b.Get(p[s], p[s])-b.Get(p[r], p[r])) +
		(a.Get(r, s)-a.Get(s, r))*(b.Get(p[s], p[r])-b.Get(p[r], p[s])))
	for i = 0; i < p.Len(); i++ {
		if i != r && i != s {
			d += (a.Get(i, r)-a.Get(i, s))*(b.Get(p[i], p[s])-b.Get(p[i], p[r])) +
				(a.Get(r, i)-a.Get(s, i))*(b.Get(p[s], p[i])-b.Get(p[r], p[i]))
		}
	}
	return d
}

func (m Matrix) Get(i int64, j int64) int64 {
	return m.A[i*m.N+j]
}

type Matrix struct {
	N int64
	A []int64
}

File: translations_only/success/go/go-gt/localSearch/gemini/Hinted/localSearch.rs

use rand::Rng;

fn local_search(a: &Matrix, b: &Matrix, p: &mut Vec<i64>, cost: &mut i64) {
    // set of moves, numbered from 0 to index
    let mut i: i64;
    let mut j: i64;
    let mut n_mov: i64 = 0;
    let n = p.len() as i64;
    let mut move_ = vec![0; (n * (n - 1) / 2) as usize];
    n_mov = 0;
    for i in 0..n - 1 {
        for j in i + 1..n {
            move_[n_mov as usize] = n * i + j;
            n_mov += 1;
        }
    }
    let mut improved = true;
    for k in 0..2 {
        improved = false;
        for i in 0..n_mov - 1 {
            move_.swap(i as usize, unif(i + 1, n_mov - 1) as usize);
        }
        for i in 0..n_mov {
            let r = move_[i as usize] / n;
            let s = move_[i as usize] % n;
            let d = delta(a, b, p, r, s);
            if d < 0 {
                *cost += d;
                p.swap(r as usize, s as usize);
                improved = true;
            }
        }
        if !improved {
            break;
        }
    }
}

fn unif(low: i64, high: i64) -> i64 {
    let mut rng = rand::thread_rng();
    low + rng.gen_range(0..high - low + 1)
}

fn delta(a: &Matrix, b: &Matrix, p: &Vec<i64>, r: i64, s: i64) -> i64 {
    let mut i: i64;
    let mut d = (a.get(r, r) - a.get(s, s)) * (b.get(p[s as usize], p[s as usize]) - b.get(p[r as usize], p[r as usize]))
        + (a.get(r, s) - a.get(s, r)) * (b.get(p[s as usize], p[r as usize]) - b.get(p[r as usize], p[s as usize]));
    for i in 0..p.len() as i64 {
        if i != r && i != s {
            d += (a.get(i, r) - a.get(i, s)) * (b.get(p[i as usize], p[s as usize]) - b.get(p[i as usize], p[r as usize]))
                + (a.get(r, i) - a.get(s, i)) * (b.get(p[s as usize], p[i as usize]) - b.get(p[r as usize], p[i as usize]));
        }
    }
    d
}

#[derive(Debug)]
struct Matrix {
    n: i64,
    a: Vec<i64>,
}

impl Matrix {
    fn get(&self, i: i64, j: i64) -> i64 {
        self.a[(i * self.n + j) as usize]
    }
}
